backtrace-library/src/main/java/backtraceio/library/metrics/BacktraceEventsHandler.java

```diff
@@ -0,0 +1,211 @@
+import backtraceio.library.logger.BacktraceLogger;
+import backtraceio.library.models.BacktraceMetricsSettings;
+import backtraceio.library.models.json.BacktraceAttributes;
+import backtraceio.library.services.BacktraceHandlerThread;
+import backtraceio.library.services.BacktraceReportSender;
+
+public abstract class BacktraceEventsHandler<T extends Event> extends Handler {
+
+    private final static transient String LOG_TAG = BacktraceEventsHandler.class.getSimpleName();
+
+    protected final BacktraceHandlerThread backtraceHandlerThread;
+
+    /**
+     * User provided custom attributes
+     */
+    protected final Map<String, Object> customAttributes;
+
+    /**
+     * The application context. We need this in our derived classes to get the BacktraceAttributes
+     */
+    protected final Context context;
+
+    /**
+     * Time between retries if metrics submission fails
+     */
+    private final int timeBetweenRetriesMillis;
+
+    /**
+     * Http client
+     */
+    protected final Api api;
+
+    /**
+     * Submission url
+     */
+    private final String submissionUrl;
+
+    /**
+     * List of events in the event queue
+     */
+    protected ConcurrentLinkedDeque<T> events = new ConcurrentLinkedDeque<T>();
+
+    /**
+     * Maximum number of events in store. If number of events in store hit the limit
+     * BacktraceMetrics instance will send data to Backtrace.
+     */
+    private int maximumNumberOfEvents = 350;
+
+    /**
+     * The application name
+     */
+    protected String application;
+
+    /**
+     * The application version
+     */
+    protected String appVersion;
+
+    /**
+     * Create BacktraceEventsHandler instance
+     *
+     * @param context
+     * @param api
+     * @param backtraceHandlerThread
+     * @param urlPrefix
+     * @param settings
+     */
+    public BacktraceEventsHandler(Context context, Map<String, Object> customAttributes,
+                                  Api api, final BacktraceHandlerThread backtraceHandlerThread, String urlPrefix, BacktraceMetricsSettings settings) {
+        // This should always have a nonnull looper because BacktraceHandlerThread starts in the
+        // constructor and getLooper blocks until the looper is ready if the thread is started
+        //
+        // We cannot cache the looper here since the super call has to go first (i.e: before
+        // declaring and assigning a Looper variable)
+        super(backtraceHandlerThread.getLooper());
+        if (backtraceHandlerThread.getLooper() == null) {
+            throw new NullPointerException("Expected nonnull looper, this should not happen");
+        }
+        this.context = context;
+        this.customAttributes = customAttributes;
+        this.backtraceHandlerThread = backtraceHandlerThread;
+        this.api = api;
+        this.submissionUrl = settings.getBaseUrl() + "/" + urlPrefix + "/submit?token=" + settings.getToken() + "&universe=" + settings.getUniverseName();
```

Comment:
add to BacktraceMetricsSettings method `getSubmissionUrl()` and move logic of constructing this url string there

Reply:
Will do, thank you!